spatial/r3/icosahedron_example_test.go

```diff
@@ -0,0 +1,104 @@
+// Copyright ©2022 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package r3_test
+
+import (
+	"fmt"
+
+	"gonum.org/v1/gonum/spatial/r3"
+)
+
+func ExampleTriangle_icosphere() {
+	// This example generates a 3D icosphere from
+	// a starting icosahedron by subdividing surfaces.
+	// See https://schneide.blog/2016/07/15/generating-an-icosphere-in-c/.
+	const subdivisions = 5
+	// vertices is a slice of r3.Vec
+	// triangles is a slice of [3]int indices
+	// referencing the vertices.
+	vertices, triangles := icosahedron()
+	for i := 0; i < subdivisions; i++ {
+		vertices, triangles = subdivide(vertices, triangles)
+	}
+	var faces []r3.Triangle
+	for _, t := range triangles {
+		var face r3.Triangle
+		for i := 0; i < 3; i++ {
+			face[i] = vertices[t[i]]
+		}
+		faces = append(faces, face)
+	}
+	fmt.Println(faces)
+	// The 3D rendering of the icosphere is left as an exercise to the reader.
+}
+
+// edgeIdx represents an edge of the icosahedron
+type edgeIdx [2]int
+
+func subdivide(vertices []r3.Vec, triangles [][3]int) ([]r3.Vec, [][3]int) {
+	// We generate a lookup table of all newly generated vertices so as to not
+	// duplicate new vertices. edgeIdx has lower index first.
+	lookup := make(map[edgeIdx]int)
+	var result [][3]int
+	for _, triangle := range triangles {
+		var mid [3]int
+		for edge := 0; edge < 3; edge++ {
+			mid[edge], vertices = subdivideEdge(lookup, vertices, triangle[edge], triangle[(edge+1)%3])
+		}
+		newTriangles := [][3]int{
+			{triangle[0], mid[0], mid[2]},
+			{triangle[1], mid[1], mid[0]},
+			{triangle[2], mid[2], mid[1]},
+			{mid[0], mid[1], mid[2]},
+		}
+		result = append(result, newTriangles...)
+	}
+	return vertices, result
+}
+
+// subdivideEdge takes the vertices list and indices first and second which
+// refer to the edge that will be subdivided.
+// lookup is a table of all newly generated vertices from
+// previous calls to subdivideEdge so as to not duplicate vertices.
+func subdivideEdge(lookup map[edgeIdx]int, vertices []r3.Vec, first, second int) (int, []r3.Vec) {
```

Comment:
perhaps also return the lookup table, to future proof any behaviour change in the guts of the `map` implementation?

Reply:
I don't think that's necessary; maps can't change their reference value behaviour because of the compatibility promise.